### DISCUSSION TITLE
Changing contents not fired by an event in vizia

### DISCUSSION THREAD

**Discussion by Maikel1963** [vizia/vizia#446] [Questions]
Hello
I have worked through the book and so far everything works fine. But now I want to add functionality that is not originated in pressing a button or entering text. Clinging to the example of the book I would like to have a timer that increments the counter every 10 seconds.
This timer should be outside of the vizia application.
e.g should be a spawned thread like
`thread::spawn(move || loop {`
`   counter +=1;`
`   std::thread::sleep(Duration::from_millis(5000));`
`});`

But I do not know how to access counter from the outside. I do not want to use a timer event from Application

**Comment by YouKnow-sys** [comment]
You can spawn a thread using the `spawn` method on `Context`.
```rs
cx.spawn(...)
```
This will give you a `ContextProxy` that you can use to send events to Vizia main thread.
And if you don't want the default thread that `cx.spawn` give you, You can create a `ContextProxy` using the `Context::get_proxy` and move it to your own thread.
And also if all you need is a timer, you can use `Context::add_timer`.

**Reply by Maikel1963** [reply]
Thanks for the answer. No, the timer was just an example. I get data from an outside source.
Ok, when I insert  `cx.spawn(something)` in the code, something is supposed to be a target. If I want to address a routine outside of application, could I just write `cx.spawn(|cx| routine(cx))`?

**Reply by Maikel1963** [reply]
Ok, I asked to early. So far this seems to work. But how do I access "counter" now?
If I try to use `cx.AppData::counter ...` inside of "routine" that does not compile

**Reply by geom3trik** [reply]
`cx.spawn` is just like `thread::spawn` but provides a way to send events back from the spawned thread to the main thread. To use something from the main thread you'd have to either copy the value, or use an `Arc` + an atomic, or an `Arc` + `Mutex` as with ordinary threads. You'll then need to use `map()` on lenses to the `Arc` in order to react to the inner value.

**Comment by Maikel1963** [comment]
I am not familiar with Arc and/or Mutex by now. I tried:
```Rust
fn main () {
  let number = Arc::new(Mutex::new(0));
[...]
  Application::new(move |cx| {
[...]
    AppData {
      count: number.lock().unwrap().clone(),
    }.build(cx)
    cx.spawn(move |cx| routine(cx, number.clone()));
[...]
.run();

}
fn routine(cx: &mut ContextProxy, data: Arc<Mutex<i32>>) -> () {
  let mut number =data.lock().unwrap();
  *number = 15;
}

```` 
This compiles, but I can't figure out how to display the changed number.
Where do have to map() and what to write into the brackets

**Reply by Fredemus** [reply]
The count in your AppData is just a new integer with its value copied from `number` when creating the AppData. You need to store the Arc instead: `count: number.clone()`. 
And then when using it, access the integer inside the Arc: `AppData::count.map(|x| c.lock.unwrap.().clone())`
